File: SPOJ/LKS.py
S,N = map(int,raw_input().split())
val = []
wt = []

for _ in xrange(N):
	v,w = map(int,raw_input().split())
	wt.append(w)
	val.append(v)

flag = 0

ans = [[0]*(S+1) for i in xrange(2)]

# ans = [[0]*(S+1) for i in xrange(N+1)]

for i in xrange(N):
	for j in xrange(1,S+1):
		if wt[i] > j:
			ans[not flag][j] = ans[flag][j]
		else:
			ans[not flag][j] = max(val[i] + ans[flag][j-wt[i]],ans[flag][j])
	flag = not flag

print ans[flag][S]

File: SPOJ/LKS.cpp
#include <bits/stdc++.h>
#define l long
#define pb push_back

long int dp[2][2000001]
int v[12345678],w[12345678];

int main(){
	int K,N,flag=0;
	scanf("%d %d",&K,&N);
	for(int n=0;n<N;n++){
		scanf("%d %d",&v[n],&w[n]);
	}
	memset(dp,0,sizeof dp);
	for(int n=0;n<N;n++){
		for(int k=0;k<K;k++){
			if (w[n]>k){
				dp[!flag][k] = dp[flag][k];
			}
			else{
				dp[!flag][k] = max(dp[flag][k],v[n]+dp[flag][k-wt[n]])
			}
		}
	}
	printf("%ld",dp[flag][K]);
	return 0;
}